### PR TITLE
feat: Add PSM refill on swap (SC-683)

### PR DESCRIPTION
**Pull request by lucas-manuel**
NOTE: Had to update the fork block to get higher debt cieling params for the new DSS lite PSM on mainnet so updated some CCTP nonces.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -19,3 +19,6 @@
 [submodule "lib/sdai"]
 	path = lib/sdai
 	url = https://github.com/makerdao/sdai
+[submodule "lib/dss-lite-psm"]
```
Comment (hexonaut) [inline .gitmodules:22]: I actually wouldn't import this dependency in this case. I would just use an inline interface since none exists in this repo. You shouldn't be using the implementation for an interface on a fork.

File: src/MainnetController.sol [hunk 2]
```diff
@@ -282,30 +284,51 @@ contract MainnetController is AccessControl {
...
+
+                uint256 swapAmount = remainingDaiToSwap < psmDaiBalance
+                    ? remainingDaiToSwap
+                    : psmDaiBalance;
+
+                _swapUSDCToDAI(swapAmount / psmTo18ConversionFactor);
+
+                remainingDaiToSwap -= swapAmount;
+            }
+        }
```
Comment (hexonaut) [inline src/MainnetController.sol:320]: This can be written more succinctly:

```
 	// If amount is larger than limit it must be split into multiple calls
        uint256 limit = dai.balanceOf(address(psm)) / psmTo18ConversionFactor;

        // Edge case where PSM is empty
        if (limit == 0) {
        	psm.fill();
        	limit = dai.balanceOf(address(psm)) / psmTo18ConversionFactor;
        }

        while (usdcAmount > limit) {
            _swapUSDCToDAI(limit);
            usdcAmount -= limit;
            psm.fill();
            limit = dai.balanceOf(address(psm)) / psmTo18ConversionFactor;
        }

        // Swap remaining amount (if any)
        if (usdcAmount > 0) {
            _swapUSDCToDAI(usdcAmount);
        }
```

**Review by hexonaut — CHANGES_REQUESTED** [review]
Some changes with the logic. Will review the tests after updates.

File: .gitmodules (hunk 1 above)
Comment (lucas-manuel) [reply .gitmodules]: Removed

File: src/MainnetController.sol (hunk 2 above)
Comment (lucas-manuel) [reply src/MainnetController.sol:320]: Discussed on call, added this and then opted for original implementation. Its less succinct but handles edge cases inherently and is easier to test

**Review by hexonaut — CHANGES_REQUESTED** [review]
Some minor changes, but looks good.

File: test/mainnet-fork/PsmCalls.t.sol [hunk 3]
```diff
@@ -127,10 +127,57 @@ contract MainnetControllerSwapUSDCToUSDSFailureTests is ForkTestBase {
...
+
+        assertEq(maxSwapAmount, 450_281_089.262716e6);
+
+        deal(address(usdc), address(almProxy), maxSwapAmount + 1);
+
+        vm.startPrank(relayer);
+        vm.expectRevert("DssLitePsm/nothing-to-fill");
+        mainnetController.swapUSDCToUSDS(maxSwapAmount + 1);
+
+        mainnetController.swapUSDCToUSDS(maxSwapAmount);
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply test/mainnet-fork/PsmCalls.t.sol:164]: Added

File: test/mainnet-fork/PsmCalls.t.sol [hunk 4]
```diff
@@ -170,6 +217,165 @@ contract MainnetControllerSwapUSDCToUSDSTests is ForkTestBase {
...
+
+        // NOTE: Art == dai here because rate is 1 for PSM ilk
+        ( uint256 Art,,,, ) = dss.vat.ilks(PSM_ILK);
+
+        assertEq(Art, 1_878_319_962.52582e18);
+        assertEq(Art, 2_000_000_000e18 - fillAmount);
+
+        vm.prank(relayer);
+        // vm.expectEmit(PSM);
+        // emit Fill(fillAmount);
```
[the remark this answers is not in the document]
Comment (lucas-manuel) [reply test/mainnet-fork/PsmCalls.t.sol:271]: Wasn't supposed to be sorry, added back

**Review by hexonaut — APPROVED** [review]
